master/internal/resourcemanagers/agent/agent_state.go

```diff
@@ -378,3 +414,274 @@ func (a *AgentState) patchSlotState(
+			enabled: slotEnabled{
+				deviceAdded:  true,
+				agentEnabled: as.UserEnabled,  // TODO
+				userEnabled:  as.UserEnabled,  // TODO
+				draining:     as.UserDraining, // TODO
+			},
+		}
+		if sd.ContainerID != nil {
+			devices[sd.Device] = sd.ContainerID
+		} else {
+			devices[sd.Device] = nil
+		}
+	}
+
+	containerState := make(map[cproto.ID]*cproto.Container)
+
+	if len(as.Containers) > 0 {
+		containerSnapshots := make([]ContainerSnapshot, 0, len(as.Containers))
+		err := db.Bun().NewSelect().Model(&containerSnapshots).
+			Where("container_id IN (?)", bun.In(as.Containers)).
+			Scan(context.TODO())
+		if err != nil {
+			return nil, err
+		}
+
+		for _, containerSnapshot := range containerSnapshots {
+			container := containerSnapshot.ToContainer()
+			containerState[container.ID] = &container
+		}
+	}
+
+	result := AgentState{
+		maxZeroSlotContainers: as.MaxZeroSlotContainers,
+		resourcePoolName:      as.ResourcePoolName,
+		Label:                 as.Label,
+		uuid:                  parsedUUID,
+		enabled:               as.UserEnabled,
+		draining:              as.UserDraining,
+		slotStates:            slotStates,
+		Devices:               devices,
+		containerAllocation:   make(map[cproto.ID]*actor.Ref),
+		containerState:        containerState,
+	}
+
+	return &result, nil
+}
+
+func (a *AgentState) restoreContainersField() error {
+	containerIDs := maps.Keys(a.containerState)
+
+	c2a, err := loadContainersToAllocationIds(containerIDs)
+	if err != nil {
+		return err
+	}
+
+	containers := make(map[cproto.ID]*actor.Ref)
+	for contID := range c2a {
+		ref := task.GetAllocation(c2a[contID])
+		if ref != nil {
+			containers[contID] = ref
+		}
+	}
+	log.WithField("agent-id", a.string()).Debugf("restored containers: %d", len(containers))
+
+	maps.Copy(a.containerAllocation, containers)
+
+	return nil
+}
+
+func clearAgentStates(agentIds []AgentID) error {
+	_, err := db.Bun().NewDelete().Where("agent_id in (?)", agentIds).Exec(context.TODO())
+
+	return err
+}
+
+func updateContainerState(c *cproto.Container) error {
+	snapshot := NewContainerSnapshot(c)
+	_, err := db.Bun().NewUpdate().Model(&snapshot).
+		Where("container_id = ?", snapshot.ID).
+		Column("state", "devices").
+		Exec(context.TODO())
+
+	return err
+}
+
+func loadContainersToAllocationIds(
+	containerIDs []cproto.ID) (map[cproto.ID]model.AllocationID, error) {
+	cs := []ContainerSnapshot{}
+	result := []map[string]interface{}{}
+	rr := map[cproto.ID]model.AllocationID{}
+
+	if len(containerIDs) == 0 {
+		return rr, nil
+	}
+
+	err := db.Bun().NewSelect().Model(&cs).
+		Join("JOIN allocation_resources al_res ON al_res.resource_id = rmac.resource_id").
+		Where("container_id IN (?)", bun.In(containerIDs)).
+		Column("container_id", "allocation_id").
+		Scan(context.TODO(), &result)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, row := range result {
+		rr[cproto.ID(row["container_id"].(string))] = model.AllocationID(row["allocation_id"].(string))
```

Comment:
why this way?

Reply:
we're selecting two columns from joined tables when there's no single struct for them, and ad-hoc bun Scan into `[]map[string]interface{}{}` seemed to be the way to do that. do you think there's a better way?